### PR TITLE
feat: Detect platform-side inference errors

### PR DESCRIPTION
**Pull request by statxc**
## Detect platform-side inference errors so agents aren't penalized for provider failures

Closes #331

### Problem

When an AI provider goes down or returns server errors (500, 502, etc.), the agent's `inference()` calls return `None`. The agent keeps running but produces a bad or empty patch because it has no LLM to work with. The platform then scores this patch normally - the agent gets a 0 for something that wasn't its fault.

There was no mechanism to distinguish "the agent wrote bad code" from "the providers were broken."

### Solution

Track platform-side inference errors per evaluation run and flag the run as a platform error when the count exceeds a configurable threshold.

**Platform errors** are provider failures that the agent can't control:
- `500` Internal Server Error
- `502` Bad Gateway
- `503` Service Unavailable
- `504` Gateway Timeout
- `-1` Internal provider error

**Non-platform errors** (400, 404, 422, 429) are excluded - those are the agent's fault (bad request, wrong model, exceeded cost limit).

### What changed

| File | What changed |
|------|-------------|
| `inference_gateway/error_hash_map.py` **(new)** | `ErrorHashMap` class that tracks inference error counts per `evaluation_run_id`, with the same auto-cleanup pattern as the existing `CostHashMap`. |
| `inference_gateway/config.py` | Added `MAX_INFERENCE_ERRORS_PER_EVALUATION_RUN` (defaults to 5 if not set in `.env`). Existing deployments won't break. |
| `inference_gateway/main.py` | Counts platform errors after each inference/embedding call. Blocks further requests with `503` once the threshold is hit. Extended `/api/usage` to include `inference_errors` and `max_inference_errors`. Added `logger.warning()` when errors are counted and when threshold blocks a request. |
| `models/evaluation_run.py` | Added `PLATFORM_TOO_MANY_INFERENCE_ERRORS = 3050` in the 3xxx platform error range. |
| `validator/main.py` | After agent finishes, queries `/api/usage` on the inference gateway with a 10s timeout. If errors exceed the limit, marks the run as a platform error (3050) instead of scoring the patch. Also wired up the `extra` field in `EvaluationRunException` handling - it was designed but never passed through. Now `agent_logs` are included when reporting platform errors. |
| `tests/test_inference_error_tracking.py` **(new)** | 19 tests covering `ErrorHashMap` unit behavior, platform error classification, error code validation, and integration tests against both inference and embedding gateway endpoints. |

### How it works end-to-end

```
Agent calls inference() → provider returns 500 → gateway counts error → agent gets None
Agent calls inference() → provider returns 500 → gateway counts error → agent gets None
...5th error...
Agent calls inference() → gateway returns 503 (blocked) → agent finishes with bad patch
Validator checks /api/usage → sees errors >= limit → marks run as PLATFORM error (3050)
→ Agent is not scored on this run
```

### Config

Add to your `.env` if you want to override the default:
```
MAX_INFERENCE_ERRORS_PER_EVALUATION_RUN=5
```

### Testing

```bash
python3 -m pytest tests/test_inference_error_tracking.py -v
```

### CODE REVIEW

**Review by camfairchild — CHANGES_REQUESTED** [review]

**Review by camfairchild — APPROVED** [review]

**Comment by statxc** [comment]
@ibraheem-abe Sorry to bother you again. I'd appreciate you approve this PR if it has no problem.

**Comment by ibraheem-abe** [comment]
@statxc 
I would like to test and review this with a working dev environment to ensure we don't miss or induce any edge cases. 
Thank you for your patience!

File: inference_gateway/error_hash_map.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self):
+        self.error_hash_map = {}
+        self.last_cleanup_at = time.time()
+
+
+
+    def _cleanup(self):
+        now = time.time()
+        if now - self.last_cleanup_at > ERROR_HASH_MAP_CLEANUP_INTERVAL_SECONDS:
+            self.error_hash_map = {k: v for k, v in self.error_hash_map.items() if now - v.last_accessed_at < ERROR_HASH_MAP_CLEANUP_INTERVAL_SECONDS}
```
Comment (CryptoAardvark) [inline inference_gateway/error_hash_map.py]: Suggestion(non-blocking): since this map is accessed from concurrent FastAPI requests, the dict reassignment here (and the check-then-increment pattern in `main.py`) might benefit from a `threading.Lock`. Concurrent requests for the same `evaluation_run_id` could slip past the threshold check before blocking kicks in.

File: inference_gateway/error_hash_map.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            self.last_cleanup_at = now
+
+
+
+    def get_inference_errors(self, uuid: UUID) -> int:
+        self._cleanup()
+
+        if uuid in self.error_hash_map:
+            self.error_hash_map[uuid].last_accessed_at = time.time()
```
Comment (CryptoAardvark) [inline inference_gateway/error_hash_map.py]: Suggetion(non-blocking): `get_inference_errors` refreshes `last_accessed_at` on reads. If `/api/usage` polls this for a completed run, it could keep the entry alive past the 60s **TTL**. I think this might be worth moving the touch out of the `get_*` path so reads stay pure.

**Review by CryptoAardvark — COMMENTED** [review]
Hi @statxc, hope you don't mind a drive-by review, this PR caught my eye and the design is really thoughtful, especially the separation of platform errors from agent-caused ones and the `-1` sentinel for network failures. I had two small observations I wanted to share in case they're useful. Please feel free to disregard anything that doesn't fit the project's conventions. 🙏

File: inference_gateway/error_hash_map.py (hunk 1 above)
Comment (statxc) [reply inference_gateway/error_hash_map.py]: Good catch on the race condition. The fix would actually need to be an `asyncio.Lock` in the request handler (around the check-await-increment sequence in `main.py`), not a `threading.Lock` in the hash map.  
The individual methods are already atomic in asyncio. Same pattern exists in `CostHashMap` for cost limits. Worst case is one extra error slipping past the threshold, so I'll leave this as-is for now.

File: inference_gateway/error_hash_map.py (hunk 2 above)
Comment (statxc) [reply inference_gateway/error_hash_map.py]: Fixed. thanks

**Review by CryptoAardvark — APPROVED** [review]

**Comment by statxc** [comment]
@camfairchild Sorry for ping again. But it has been took long time since I submitted this PR. Please give me any updates for me. thanks

**Comment by camfairchild** [comment]
> @camfairchild Sorry for ping again. But it has been took long time since I submitted this PR. Please give me any updates for me. thanks

Thanks for your interest in contributing and apologies for the delay in our response.

We've gone a different route during the refactor so unfortunately the PR isn't relevant anymore. But still very appreciated.

I look forward to future contributions of yours after the refactor is merged.
